cf/src/main/scala/typeformation/cf/Encoding.scala

```diff
@@ -256,102 +278,170 @@ object Encoding {
+      def f = emptyToNull _
 
-    Json.obj(
-      c.logicalId -> Json.obj(
-        "commands" -> f(fold(c.commands)),
-        "files" -> f(fold(c.files)),
-        "groups" -> f(fold(c.groups)),
-        "packages" ->  pkgsByFormat.asJson,
-        "services" -> c.services.asJson,
-        "sources" -> f(fold(c.sources)),
-        "users" -> f(fold(c.users))
+      Json.obj(
+        c.logicalId -> Json.obj(
+          "commands" -> f(foldKeyVals(c.commands)),
+          "files" -> f(foldKeyVals(c.files)),
+          "groups" -> f(foldKeyVals(c.groups)),
+          "packages" -> pkgsByFormat.asJson,
+          "services" -> c.services.asJson,
+          "sources" -> f(foldKeyVals(c.sources)),
+          "users" -> f(foldKeyVals(c.users))
+        )
       )
-    )
   }
 
   implicit val cfnInitEncoder: Encoder[Init] = Encoder.instance { i =>
     Json.obj(
-      "AWS::CloudFormation::Init" -> Json.obj(
-        "configSets" -> i.configSets.asJson
-      ).deepMerge(fold(i.configs.toList))
+      "AWS::CloudFormation::Init" -> Json
+        .obj(
+          "configSets" -> i.configSets.asJson
+        )
+        .deepMerge(foldKeyVals(i.configs.toList))
     )
   }
 
-  private [cf] def emptyToNull(o: Json): Json = {
+  implicit val iamActionEncoder: Encoder[iam.Action] =
+    implicitly[Encoder[Seq[String]]].contramap(_.value)
+
+  implicit val encodeIamPrincipal: Encoder[iam.Principal] = {
+    import iam.Principal._
+    Encoder.instance {
+      case Aws(arns @ _*)       => Json.obj("AWS" -> unwrapSingleton(arns))
+      case Service(values @ _*) => Json.obj("Service" -> values.asJson)
+      case CanonicalUser(value) => Json.obj("CanonicalUser" -> value.asJson)
+      case Federated(value)     => Json.obj("Federated" -> value.asJson)
+    }
+  }
+
+  implicit val encodeIamCondition: Encoder[iam.Condition] =
+    Encoder.instance { c =>
+      val suffix = if (c.hasIfExists) "IfExists" else ""
+      val labelWithModifiers =
+        c.quantifier.fold[String](c.label)(q => s"${q.id}:${c.label}") ++ suffix
+
+      Json.obj(
+        labelWithModifiers -> Json.obj(
+          c.key.value -> unwrapSingleton(c.expressionEncoder(c.expected)))
+      )
+    }
+
+  implicit val encodeStatement: Encoder[iam.Statement] =
+    Encoder.instance { s =>
+      def invertibleKV[A: Encoder](key: String,
+                                   inv: iam.Invertible[A]): (String, Json) =
+        (if (inv.isPositive) key else s"Not$key") -> unwrapSingleton(
+          inv.value.asJson)
+
+      Json.obj(
+        Seq(
+          "Sid" -> s.Sid.asJson,
+          "Effect" -> Json.fromString(iam.Effect.effectEnum.encode(s.Effect)),
+          invertibleKV("Resource", s.Resource),
+          invertibleKV("Action", s.Action),
+          "Condition" -> foldKeyVals(s.Condition)
+        ) ++ s.Principal
+          .map(p => Seq(invertibleKV("Principal", p)))
+          .getOrElse(Nil): _*)
+    }
+
+  implicit val encodePolicy: Encoder[iam.Policy] =
```

Comment:
why no derivation?

Reply:
well spotted... most of them need to be written by hand, so I went on autopilot